
<center><h1 style="color:blue;">Waze Churn Analysis</h1></center>

### Overview

Waze is subsidiary company of Google that provides satellite navigation on devices that supports GPS. The leadership at Waze has tasked the data team with developing a machine learning model for anticipating user attrition. This model relies on information gathered from individuals using the Waze application. 

### Objective

I have been assigned the responsibility to develop a machine learning model by the Waze leadership. The objective is to predict user churn using data obtained from users of the Waze app.

To develop a ML model for predicting user churn using data obtained from users of the Waze app

<h2 style ="color:blue">Stage 1: Inspect and Analyze Data</h2>

### Purpose

To investigate and understand the data by constructing dataframe with python and perform cursory inspection of the dataset

### Key Activities

* Prepae the data by reviewing the data dictionary, and conducting an initial exploration of the dataset to pinpoint essential variables for the stakeholder's needs
* Generate a pandas dataframe to facilitate data learning, future exploratory data analysis (EDA), and statistical tasks

* Aggregate comprehensive summary details about the dataset to guide subsequent actions

### **Activity 1a. Imports and data loading**

In [1]:
# Import packages for data manipulation
import pandas as pd
import numpy as np

In [2]:
# Load dataset into dataframe
# Load dataset into dataframe
df = pd.read_csv(r'E:/Advance Google Data Analytics/Waze/waze_dataset.csv')
df.head(10)


ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

### **Activity 1b. Information Summary**

In [3]:
df.head(10)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


**Key Takeaways:**

1. First 10 observations have no missing value, in-depth analysis for missing value still needed.
2. The dataset consists of 14,999 rows and 13 columns. The variable types are as follows:
    * label and device are categorical variables of type object.
    * total_sessions, driven_km_drives, and duration_minutes_drives are numerical variables of type float64.
    * The remaining variables are numerical variables of type int64.
3. Within the dataset, the label column contains 700 missing values.

### **Activity 1b. Null values and summary statistics**
Contrast the summary statistics of the 700 rows that are missing labels with the summary statistics of the rows containing complete data.


In [5]:
null_df = df[df['label'].isnull()]
null_df.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [6]:
# Isolate rows without null values
not_null_df = df[~df['label'].isnull()]
# Display summary stats of rows without null values
not_null_df.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

**Key takeaway:**

>  The means and standard deviations are fairly consistent between the two groups having missing data and complete data

### **Activity 1c. Device Count by Null Values**

In [7]:
# Get count of null values by device
null_df['device'].value_counts()

# Calculate % of iPhone nulls and Android nulls
null_df['device'].value_counts(normalize=True)

# Calculate % of iPhone users and Android users in full dataset
df['device'].value_counts(normalize=True)

iPhone     0.644843
Android    0.355157
Name: device, dtype: float64

**Key takeaway:**

>  Out of **700** rows with missing values, **447** were iPhone users and **253** were Android users.

### **Activity 1d. Device Ratio Comparison**

Comparing device ratio within the **Null Values** in dataset

In [8]:
# Calculate % of iPhone nulls and Android nulls
null_df['device'].value_counts(normalize=True)

iPhone     0.638571
Android    0.361429
Name: device, dtype: float64

Comparing to the device ratio within **Complete** dataset

In [9]:
# Calculate % of iPhone users and Android users in full dataset
df['device'].value_counts(normalize=True)

iPhone     0.644843
Android    0.355157
Name: device, dtype: float64

**Key takeaway:**

>  * The percentage of missing values by each device is consistent with their representation in the data overall.
> * There is nothing to suggest a non-random cause of the missing data

### **Activity 1d. Churned Vs Retained Users**

In [10]:
# Calculate counts of churned vs. retained
print(df['label'].value_counts())
print()
print(df['label'].value_counts(normalize=True))

retained    11763
churned      2536
Name: label, dtype: int64

retained    0.822645
churned     0.177355
Name: label, dtype: float64


### KeyTakeaway:
> This dataset contains 82% retained users and 18% churned users.

Median comparison of retained and churned Users, we use median to minimize outliers unduly affect the analysis

In [11]:
# Calculate median values of all columns for churned and retained users
df.groupby('label').median(numeric_only=True)

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

### KeyTakeaway:
> * Churned users had about ~3 more drives on average in the past month than retained users. Meanwhile, retained users used the app for more than twice the number of days compared to churned users in the same period.

> * The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

> * This pattern indicates that churned users engaged in a higher frequency of drives within a shorter span, with farther trips and longer durations.  This could potentially hint at distinctive user profiles

### **Activity 1e. Comparison - Average Kilometers/Drive**

In [12]:
# Group data by `label` and calculate the medians
medians_by_label = df.groupby('label').median(numeric_only=True)
print('Median kilometers per drive:')
# Divide the median distance by median number of drives
medians_by_label['driven_km_drives'] / medians_by_label['drives']

Median kilometers per drive:


label
churned     73.053113
retained    73.716694
dtype: float64

### KeyTakeaway:
> The median user from both groups drove ~73 km/drive. How many kilometers per driving day was this?

**Median kilometers per driving day**

In [13]:
  # Divide the median distance by median number of driving days
print('Median kilometers per driving day:')
medians_by_label['driven_km_drives'] / medians_by_label['driving_days']

Median kilometers per driving day:


label
churned     608.775944
retained    247.477472
dtype: float64

**Median number of drives per driving day for each group**

In [14]:
# Divide the median number of drives by median number of driving days
print('Median drives per driving day:')
medians_by_label['drives'] / medians_by_label['driving_days']

Median drives per driving day:


label
churned     8.333333
retained    3.357143
dtype: float64

### KeyTakeaway:
> * Median churned user covered **608 kilometers** per driving day last month, nearly **250%** more than retained users. Similarly, the median churned user had significantly more drives per drive day compared to retained users

> * These figures clearly show that the users in the dataset, regardless of churn status, are avid drivers. It's likely that the data doesn't represent typical drivers. The sample of churned users might consist largely of **long-haul truckers**

> * Given the extensive driving habits of these users, it's advisable for Waze to consider collecting additional data on these highly active drivers. The underlying reason for their high mileage could potentially shed light on why the Waze app might not align with their unique requirements. These needs might differ significantly from those of a more usual driver, like a daily commuter

**Counts of each device type for each group - churned and retained**

In [15]:
df.groupby(['label', 'device']).size()

label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

Now, within each group, churned and retained, calculate what percent was Android and what percent was iPhone.

In [16]:
# For each label, calculate the percentage of Android users and iPhone users
df.groupby('label')['device'].value_counts(normalize=True)

df.groupby('label')['device'].value_counts(normalize = True)

label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: device, dtype: float64

### Key Takeaway:
> * The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.

<h2 style="color:blue;">Conclusion - Inspect and Analyze Data</h2>

> * The dataset has 700 missing values in the `label` column. There was no obvious pattern to the missing values
> * Mean is subject to the influence of outliers, while the median represents the middle value of the distribution regardless of any outlying values
> * investigation gave rise to further questions, the median user who churned drove 608 kilometers each day they drove last month, which is almost 250% the per-drive-day distance of retained users. It would be helpful to know how this data was collected and if it represents a non-random sample of users
> * Android users comprised approximately 36% of the sample, while iPhone users made up about 64%
> * Generally, users who churned drove farther and longer in fewer days than retained users. They also used the app about half as many times as retained users over the same period
> * The churn rate for both iPhone and Android users was within one percentage point of each other. There is nothing suggestive of churn being correlated with device